add feature to report bad comments

nif33 opened the pull request:
add a report button on every comment
Upon clicking button, a fetch is made to the database to increase the number of reports by one
if a comment is reported twice it will not be shown to users

nif33 committed:
add feature to report bad comments

add a report button on every comment
if a comment is reported twice it will not be shown to users

nif33 committed:
add feature to report bad comments

add a report button on every comment
if a comment is reported twice it will not be shown to users

juanalonsocruz commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -50,13 +52,17 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    Query query = new Query("Comment");
+    query.addSort("numReports", SortDirection.ASCENDING); // required sorting to apply filter
+    query.addSort("timestamp", SortDirection.DESCENDING); // display in order of recency
+    query = query.addFilter("numReports", Query.FilterOperator.LESS_THAN, 2); // only show comments with less than 2 reports
```
Comments are redundant because the code is self explanatory. Check out the `To Comment or Not to Comment?` ToTT.

juanalonsocruz commented on portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java:
```diff
@@ -0,0 +1,37 @@
+    String id = request.getParameter("id");
+    try {
+      Key commentKey = KeyFactory.stringToKey(id);
+  		Entity commentEntity = datastore.get(commentKey);
```
Bad indentation.

nif33 replied:
weird. Looked normal in Atom, must be some setting difference in tab spacing. I'll make sure to check the Github display next time.

juanalonsocruz commented on portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java:
```diff
@@ -0,0 +1,37 @@
+      numReports++;
+      commentEntity.setProperty("numReports", numReports);
+      datastore.put(commentEntity);
+  	} catch (EntityNotFoundException e) {
```
Minimize code inside try block. The first two lines should be enough.

juanalonsocruz commented on portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java:
```diff
@@ -0,0 +1,37 @@
+    try {
+      Key commentKey = KeyFactory.stringToKey(id);
+  		Entity commentEntity = datastore.get(commentKey);
+      long numReports = (long) commentEntity.getProperty("numReports");
+      numReports++;
+      commentEntity.setProperty("numReports", numReports);
+      datastore.put(commentEntity);
```
Andrew in a different PR. Replace `like` with `report`
> What happens if two people have your website open and like the same comment at the exact same time? ;)

nif33 replied:
I actually talked about this yesterday with Zoe. I feel like it's less important for my use case because I am counting report numbers and I arbitrarily chose 2 reports to be the number required for a comment to be taken down (it'd have to be more if the website had more traffic).
Nonetheless, I have amended my code according to this documentation: https://cloud.google.com/appengine/docs/standard/java/datastore/transactions#using_transactions
1) I accidentally amended onto the wrong commit (oops, forgot I did a merge, will make sure to check next time)
2) the documentation code uses an infinite loop and breaks from it, in class I've always learned that using infinite loops is bad practice, what is the truth? I feel like it might be less readable if I did numRetries in a for loop...

juanalonsocruz replied:
In general, you want to have accurate data for future use cases. What if in the future you want to create a new page that displayed metrics that depend directly on the number of reports?

For your (2) comment, I agree. for loop is good. I would also put everything that inside the loop into a method called something like `incrementReportCount(...)`

Overall: Please don't spend much more time on this. It is mainly a good thing to think about for capstone.

juanalonsocruz commented on portfolio/src/main/webapp/script.js:
```diff
@@ -35,18 +35,31 @@ function newCommentBox(comment){
   commentText.innerText = comment.text;
+  commentReport.innerText = "Report";
+  commentReport.onclick = function() {
+    addReport(comment.id);
```
Nit: In the back-end you are adding reports but in the front-end to the users they are reporting a comment. I recommend renaming to reportComment or something along those lines.

nif33 committed:
fix merge conflicts

nif33 committed:
fix merge conflicts

nif33 committed:
fix merge conflicts

nif33 committed:
Merge branch 'JenniferDevelopment' of https://github.com/nif33/my-portfolio into JenniferDevelopment

nif33 committed:
Merge branch 'JenniferDevelopment' of https://github.com/nif33/my-portfolio into JenniferDevelopment

nif33 committed:
Merge branch 'JenniferDevelopment' of https://github.com/nif33/my-portfolio into JenniferDevelopment

juanalonsocruz commented on portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java:
```diff
@@ -0,0 +1,58 @@
+          String id = request.getParameter("id");
+          Key commentKey;
+          Entity commentEntity;
+          try {
+            commentKey = KeyFactory.stringToKey(id);
+            commentEntity = datastore.get(commentKey);
+        	} catch (EntityNotFoundException e) {
+        		throw new RuntimeException("Key is not an entity in the database");
+        	}
```
Indentation. Wonder whats happening with Atom and github

juanalonsocruz replied:
No need to catch and throw a similar exception. `e` possibly has more information in its message about what went wrong.

nif33 committed:
move report increment to a function

change retry counter loop into a for loop
move report increment into a function called by the loop
indentation fixes

nif33 commented on portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java:
```diff
@@ -22,37 +22,39 @@
+        numReports++;
+        commentEntity.setProperty("numReports", numReports);
+        datastore.put(transaction, commentEntity);
+        transaction.commit();
```
I moved this back into the try/catch because it was giving me errors related to the commentEntity might not be initiated. I will go back to this issue at the end of the week if I have time so I don't spend more time on this part of the project right now.

nif33 commented on portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java:
```diff
@@ -22,37 +22,39 @@
-          }
-        }
+    while(numRetries > 0) {
+      numRetries = incrementReportCount(id, numRetries);
```
I used a while loop instead of a for because the function has to return numRetries, and it seemed redundant to update the numRetries in a for loop

juanalonsocruz replied:
You could instead move the try catches outside the method into where the loop is. This way the helper method will be independent of retries.

nif33 committed:
remove repeated function calls

nif33 committed:
Merge branch 'master' of https://github.com/nif33/my-portfolio into JenniferDevelopment

nif33 committed:
remove duplicate functions

I think it got added during an automatic merge

nif33 committed:
Merge branch 'JenniferDevelopment' of https://github.com/nif33/my-portfolio into JenniferDevelopment

juanalonsocruz commented on portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java:
```diff
@@ -22,37 +22,39 @@
+  public int incrementReportCount(String id, int numRetries) {
+    Transaction transaction = datastore.beginTransaction();
+    try {
+      try {
```
No need to have cascading try blocks. Java lets you catch different exceptions like this

```
try {
  ...
} catch (Exception1 e) {
  ...
} catch (Exception2 e) {
  ...
} finally {
  ...
}
```

juanalonsocruz commented on portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java:
```diff
@@ -22,37 +22,39 @@
+        datastore.put(transaction, commentEntity);
+        transaction.commit();
+      } catch (EntityNotFoundException e) {
+        System.err.println("Key is not an entity in the database");
```
You should also inform the API client that something went wrong. Look into HTTPServletResponse.sendError(.,.)

juanalonsocruz commented on portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java:
```diff
@@ -22,37 +22,39 @@
+      return 0;
+    } catch (ConcurrentModificationException e) {
+      if(numRetries == 0) {
+        throw e;
```
Instead of throwing e back. You should call HTTPServletResponse.sendError(.,.).

juanalonsocruz commented on portfolio/src/main/webapp/script.js:
```diff
@@ -127,32 +127,6 @@ function addComment() {
   });
 }
 
-function verifyDelete() {
-  var inputWord = prompt('What\'s the magic word?');
-  if (inputWord == 'please') {
-    alert('Comments have been deleted.')
-    return true;
-  }
-  alert('That word is not magic.');
-  return false;
-}
-
-/**
- * Deletes all the comments from the page
- */
- function deleteComments() {
-   if (!verifyDelete()) {
-      return;
-   }
-
-   const commentContainer = document.getElementById('comment-container');
-   const request = new Request('/delete-data', {method: 'POST'});
-
-   fetch(request).then(() => {
-     clearChildren(commentContainer);
-   });
- }
-
```
Do you intend to remove the delete all functionality? If so, this should be in its own PR.

nif33 replied:
No, it was just some duplicate code that appeared, I think it added automatically during some merge request.
The file is getting kind of long and hard to manage, am I allowed to add new js files to separate things?

juanalonsocruz replied:
Yes! Anything to make it easier to read.

nif33 committed:
refactor try/catch block and transaction retries loop

juanalonsocruz commented on portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java:
```diff
@@ -22,39 +22,39 @@
-      }
+  public boolean incrementReportCount(String id, Transaction transaction) throws EntityNotFoundException, ConcurrentModificationException {
+    Key commentKey = KeyFactory.stringToKey(id);
+    Entity commentEntity = datastore.get(commentKey);
+    if(commentEntity == null) {
```
get doesn't return null according to the doc. It would throw an exception. No need to do this check or have this method return a boolean since its not being used in the caller.
https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/DatastoreService.html#get-com.google.appengine.api.datastore.Key-

Also, the transaction is only grouping the put. You want to group both the get and write. Use get(transaction, commentKey) instead.
https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/DatastoreService.html#get-com.google.appengine.api.datastore.Transaction-com.google.appengine.api.datastore.Key-

juanalonsocruz reviewed (approved):
Small changes needed but otherwise approve =D

nif33 committed:
change function to void type and add transaction to Entity.get
